src/app/components/wrapper-components/part-creation/part-creation.component.ts

```diff
@@ -294,112 +310,130 @@ export class PartCreationComponent implements OnInit, OnDestroy {
-        const onDocumentCreated: (foundUser: UserDocument[]) => void = async(foundUsers: UserDocument[]) => {
-            for (const user of foundUsers) {
-                if (user.data.state === 'offline') {
-                    await this.removeUserFromLobby(Utils.getNonNullable(user.data.username));
-                    Utils.handleError('OnlineGameWrapper: ' + user.data.username + ' is already offline!');
-                }
-            }
-        };
-        const onDocumentModified: (modifiedUsers: UserDocument[]) => void = async(modifiedUsers: UserDocument[]) => {
-            for (const user of modifiedUsers) {
-                if (user.data.state === 'offline') {
-                    await this.removeUserFromLobby(Utils.getNonNullable(user.data.username));
-                }
-            }
-        };
-        const onDocumentDeleted: (deletedUsers: UserDocument[]) => void = async(deletedUsers: UserDocument[]) => {
-            // This should not happen in practice, but if it does we can safely remove the user from the lobby
-            for (const user of deletedUsers) {
-                await this.removeUserFromLobby(Utils.getNonNullable(user.data.username));
-                Utils.handleError('OnlineGameWrapper: ' + user.data.username + ' was deleted (' + user.id + ')');
-            }
-        };
-        const callback: FirebaseCollectionObserver<User> =
-            new FirebaseCollectionObserver(onDocumentCreated, onDocumentModified, onDocumentDeleted);
-        for (const candidateName of joiner.candidates) {
-            if (this.candidateSubscription.get(candidateName).isAbsent()) {
-                const subscription: () => void = this.userService.observeUserByUsername(candidateName, callback);
-                this.candidateSubscription.set(candidateName, subscription);
+        for (const candidate of joiner.candidates) {
+            if (this.usersSubscriptions.get(candidate.id).isAbsent()) {
+                const onCandidateMissedHeartbeats: () => void = async() => {
+                    await this.removeUserFromLobby(Utils.getNonNullable(candidate.id));
+                };
+                this.observeUser(candidate.id, onCandidateMissedHeartbeats);
             }
         }
-        for (const oldCandidate of this.candidateSubscription.listKeys()) {
+        for (const oldCandidate of this.usersSubscriptions.listKeys()) {
             // Unsubscribe old candidates
             if (oldCandidate !== joiner.chosenPlayer) {
-                if (joiner.candidates.includes(oldCandidate) === false) {
+                if (joiner.candidates.some((minimalUser: MinimalUser) => minimalUser.id === oldCandidate) === false) {
                     this.unsubscribeFrom(oldCandidate);
                 }
             }
         }
     }
-    private removeUserFromLobby(username: string): Promise<void> {
+    public async startSendingPresenceTokensIfNotDone(): Promise<void> {
```

Comment:
if what is not done?

Reply:
Y a personne de ce nom là ici !
Sinon c'était if startSendingPresenceToken was not done :p